### PR TITLE
adding a test suite for YfirlesturAnnotations

### PR DESCRIPTION
**Pull request by Kristober**
The tests are dependant on one to run their own Yfirlestur service locally. https://github.com/mideind/Yfirlestur

### CODE REVIEW

**Comment by lumpidu** [comment]
First thing: rebase your commits and squash all of the build.yml commits: `git rebase -i origin/refactoring-yfirlestur-annotation-method-of-indexing`

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -2,11 +2,33 @@ name: Android CI
...
     steps:
       - name: Checkout
-        uses: actions/checkout@v2
+        uses: actions/checkout@v3
+      - name: Checkout Yfirlestur Repo
+        uses: actions/checkout@v3
+        with:
+          repository: mideind/Yfirlestur
+          path: yfirlestur
+          ref: dockerfile_development
```
Comment (lumpidu) [inline .github/workflows/build.yml:15]: I would base this on a commit SHA1 inside the master branch, not on the branch `dockerfile_development`. A branch can go away anytime, especially if it's a merge request branch. Whereas a SHA1 in the master branch is unlikely to go away (though not impossible).

File: app/src/main/java/org/grammatek/simacorrect/spellcheckerservice/SimaCorrectSpellCheckerService.kt [hunk 2]
```diff
@@ -77,14 +77,10 @@ class SimaCorrectSpellCheckerService : SpellCheckerService() {
...
                 try {
-                    var text = textInfos[i].text!!
+                    val text = textInfos[i].text!!
                     if (text.isBlank()) {
                         continue
                     }
-                    if (text.trim()[0].isLowerCase()) {
-                        val index = text.indexOf(text.trim())
-                        text = StringBuilder(text).also { it[index] = text.trim()[0].uppercaseChar() }.toString()
-                    }
```
Comment (lumpidu) [inline app/src/main/java/org/grammatek/simacorrect/spellcheckerservice/SimaCorrectSpellCheckerService.kt:87]: These changes should have been placed in your former merge request instead of here, because this MR seems to imply that it only adds a test suite and doesn't fix any bugs at the same time

File: app/src/main/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotation.kt [hunk 3]
```diff
@@ -185,7 +185,10 @@ class YfirlesturAnnotation(
             for (r in results) {
                 val annotations = arrayListOf<Annotations>()
                 for (annotation in r.annotations!!) {
-                    annotations.add(annotation)
+                    // Make sure to ignore capitalization for beginning of sentences
+                    if (!(annotation.code == "Z002" && annotation.start == 0)) {
+                        annotations.add(annotation)
+                    }
```
Comment (lumpidu) [inline app/src/main/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotation.kt:191]: ditto

File: app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt [hunk 4]
```diff
@@ -0,0 +1,219 @@
+package org.grammatek.simacorrect.spellcheckerservice
+
+import com.google.common.truth.Truth.assertThat
+import org.grammatek.apis.DevelopersApi
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+
+typealias AnnotationIndices = YfirlesturAnnotation.AnnotationIndices
+
```
Comment (lumpidu) [inline app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt:10]: In general, the tests would dramatically be more valuable, if we could have all raw texts in one or multiple separate file(s) and the test cases would just import this/these file(s). Then it would be easy to add new examples to the tests without touching any test source code lines.

We have e.g. error corpora that we could use to get well known error suggestions.

File: app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+        assertThat(suggestionsInfo).isEmpty()
+
+        dictionary.clear()
+        suggestionsInfo = YfirlesturAnnotation(response, text).getSuggestionsForAnnotatedWords(suggestionLimit, dictionary)
+        assertThat(suggestionsInfo).isNotEmpty()
+    }
+
+    @Test
+    fun `is suggestion limit enforced`() {
+        var suggestionLimit = 0
```
Comment (lumpidu) [inline app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt:126]: this test would benefit from much more raw text variations and increasing suggestionLimits from 0..n

File: app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+        suggestionInfo = ylAnnotation.getSuggestionsForAnnotatedWords(suggestionLimit, dictionary)
+        for (suggestions in suggestionInfo) {
+            assertThat(suggestions.suggestionsCount).isAtMost(suggestionLimit)
+        }
+    }
+
+    @Test
+    fun `multiple whitespaces don't throw off indices for annotations`() {
+        val suggestionLimit = 5
+        val text = "   Afhverju  er   það  eki  hægt"
```
Comment (lumpidu) [inline app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt:147]: What about trailing whitespaces, tabs, emojis at the end or before a trailing whitespace or after a leading whitespace ?

File: app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+        for (i in ylAnnotation.suggestionsIndices.indices) {
+            assertThat(ylAnnotation.suggestionsIndices[i].startChar).isEqualTo(startIndices[i])
+            assertThat(ylAnnotation.suggestionsIndices[i].endChar).isEqualTo(endIndices[i])
+        }
+    }
+
+    @Test
+    fun `capitalization for the first word of a sentence to be ignored by Yfirlestur`() {
+        val suggestionLimit = 5
+        val capitalizedText = "Mig langar ekki út."
```
Comment (lumpidu) [inline app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt:164]: also here: we should have many more examples of raw text

File: app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+        assertThat(capitalizedSuggestionsInfo).isEqualTo(nonCapitalizedSuggestionsInfo)
+    }
+
+    @Test
+    fun `special characters don't throw off annotation indices`() {
+        // Yfirlestur's python binary is compiled with different support than of Android.
+        // As a consequence, Yfirlestur and Android count codepoints differently i.e. some
+        // special characters have different lengths in Yfirlestur, from Android.
+        val suggestionLimit = 0
+        val text = "\uD83E\uDD76\u200B er kallt og \uD83E\uDD75 er heittt"
```
Comment (lumpidu) [inline app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt:183]: also here: this just tests a very special case, not the general functionality

File: .github/workflows/build.yml (hunk 1 above)
Comment (Kristober) [reply .github/workflows/build.yml:15]: addressed in the most recent PR's where this has been merged to master on Yfirlestur.

File: app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt (hunk 4 above)
Comment (Kristober) [reply app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt:10]: That was my first approach, I did however struggle to implement it since majority of these tests have a specific value to test for as oppose to a general one. 
The question is, is it worth while to refactor the to accomplish this?

Comment (lumpidu) [reply app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt:10]: Absolutely ! Then you can add e.g. snippets of the official annotated error corpora that exist for Icelandic and see how the underlying spell-checker performs. Also the text samples are really quite "constructed" and it's so much easier to amend real-world examples this way.

File: app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt (hunk 5 above)
Comment (Kristober) [reply app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt:126]: The only problem with that is it's hard to get more than 2 suggestions from Yfirlestur for a single annotation. Realistically Yfirlestur only gives 1 Annotation per word.

File: app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt (hunk 6 above)
Comment (Kristober) [reply app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt:147]: Trailing whitespaces aren't really any concern since they are basically ignore (as they should).
Good point with tabs, hadn't really taken that into account since its on phone but it could be from copied text or something. 
Adding a test case with emojis would be useful, especially since their slightly different to typical characters in terms of unicode code points

File: app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt (hunk 8 above)
Comment (Kristober) [reply app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt:155]: This is still a valuble test in my opinion, The name of the test might need to be renamed but this test helps determine if unicode characers contain "surrogate pairs" which throw of the character indices. This can't be changed to take data from an input file and adding multiple cases wouldn't be meaningful. 
Maybe renaming the test case would cause less confusion, renaming to something like "unicode surrogate pairs don't throw off character indices"

**Review by lumpidu — COMMENTED** [review]
Just a few nitpicks. Is it possible to get rid of the annotation warnings ?

File: app/src/test/res/test [hunk 9]
```diff
@@ -0,0 +1,42 @@
+[
```
[the remark this answers is not in the document]
Comment (Kristober) [reply app/src/test/res/test:1]: `YfirlesturAnnotationTestCases.json` sound good?

[the remark this answers is not in the document]
Comment (lumpidu) [reply app/src/test/res/test:1]: perfect :)

File: app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+
+@Serializable
+data class TestCases (
+    val input: String,
+    val suggestions: List<String>,
+)
+
+@RunWith(RobolectricTestRunner::class)
+class YfirlesturAnnotationTest {
+    private val api = DevelopersApi("http://localhost:5002")
```
[the remark this answers is not in the document]
Comment (Kristober) [reply app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt:23]: Don't we want the CI to spin up its own Yfirlestur service for testing?

[the remark this answers is not in the document]
Comment (lumpidu) [reply app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt:23]: Yes sure. But from the code alone it's not clear, how anybody on `locahost:5002` could possibly listen to the request. It would be helpful to describe the test scenario.

[the remark this answers is not in the document]
Comment (Kristober) [reply app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt:23]: Yeah that's fair, I'll have a small document for the tests in a couple of minutes for review

File: app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt [hunk 11]
```diff
@@ -20,8 +20,8 @@ data class TestCases (
 
 @RunWith(RobolectricTestRunner::class)
 class YfirlesturAnnotationTest {
-    private val api = DevelopersApi("http://localhost:5002")
-    private val file = File("src/test/res/test").readText()
+    private val api = DevelopersApi("https://yfirlestur.is")
```
Comment (lumpidu) [inline app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt:23]: shouldn't we use different url here ?

Comment (lumpidu) [reply app/src/test/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotationTest.kt:23]: otherwise doc lgtm
